src/bson_util.rs

```diff
@@ -162,13 +172,42 @@ pub(crate) fn extend_raw_document_buf(
     this: &mut RawDocumentBuf,
     other: RawDocumentBuf,
 ) -> Result<()> {
+    let mut keys: HashSet<String> = HashSet::new();
```

Comment:
I got nervous about this mostly in the context of `append_options_to_raw_document`, where it would silently generate invalid documents if there were duplicate keys, whereas `append_options` would silently overwrite the old values.  This does introduce the overhead of building the key map but I think that won't be significant, since it'll still be less overhead than constructing the body as a `Document` would have been.